src/dft.cpp

```diff
@@ -289,30 +306,52 @@ double fields::dft_norm() {
   am_now_working_on(Other);
   double sum = 0.0;
   for (int i = 0; i < num_chunks; i++)
-    if (chunks[i]->is_mine()) sum += chunks[i]->dft_norm2();
+    if (chunks[i]->is_mine()) sum += chunks[i]->dft_norm2(gv);
   finished_working();
   return std::sqrt(sum_to_all(sum));
 }
 
-double fields_chunk::dft_norm2() const {
+double fields_chunk::dft_norm2(grid_volume fgv) const {
   double sum = 0.0;
   for (dft_chunk *cur = dft_chunks; cur; cur = cur->next_in_chunk)
-    sum += cur->norm2();
+    sum += cur->norm2(fgv);
   return sum;
 }
 
 static double sqr(std::complex<realnum> x) { return (x*std::conj(x)).real(); }
 
-double dft_chunk::norm2() const {
+double dft_chunk::norm2(grid_volume fgv) const {
   if (!fc->f[c][0]) return 0.0;
   double sum = 0.0;
-  size_t idx_dft = 0;
-  const int Nomega = omega.size();
-  LOOP_OVER_IVECS(fc->gv, is, ie, idx) {
-    for (int i = 0; i < Nomega; ++i)
-        sum += sqr(dft[Nomega * idx_dft + i]);
-    idx_dft++;
+  size_t idx_dft;
+  const size_t Nomega = omega.size();
+  /* looping over chunks that have been "expanded"
+  for adjoint calculations requires some care. Namely,
+  we want to make sure we don't double count the padding
+  and can replicate results with different chunk combinations.
+  */
+  if (persist) {
+    grid_volume subgv = fgv.subvolume(is,ie,c);
+    LOOP_OVER_IVECS(fgv, is_old, ie_old, idx) {
+      idx_dft = subgv.index(c,fgv.iloc(c,idx));
+      if ((idx_dft < 0) || idx_dft > N) continue;
+      for (size_t i = 0; i < Nomega; ++i)
+          sum += sqr(dft[Nomega * idx_dft + i]);
+    } 
+  } 
+  /* note we place the if outside of the
+  loop to avoid branching. This routine gets
+  called a lot, so let's try to stay efficient
+  (at the expense of uglier code).
+   */
+  else{
+    LOOP_OVER_IVECS(fgv, is, ie, idx) {
```

Comment:
Similar to above.

Reply:
Actually it's okay because we use `IVEC_LOOP_COUNTER` here … so you don't need `subgv` at all.